allennlp/interpret/influence_interpreters/simple_influence.py

```diff
@@ -0,0 +1,352 @@
+                # we do the first .backward() and we store those used parameters
+                test_loss.backward(retain_graph=True)
+                self._used_params = [
+                    p
+                    for _, p in self.model.named_parameters()
+                    if p.requires_grad and p.grad is not None
+                ]
+                self._used_params_name = [
+                    n
+                    for n, p in self.model.named_parameters()
+                    if p.requires_grad and p.grad is not None
+                ]
+
+            # list of (parameters) gradients w.r.t. the test loss
+            test_grads: List[Tensor] = autograd.grad(test_loss, self._used_params)
+            assert len(test_grads) == len(self._used_params)
+
+            # Approximate the inverse of Hessian-Vector Product through LiSSA algorithm
+            if isinstance(self.recur_depth, float):
+                recursion_depth = int(
+                    len(self._lissa_dataloader)
+                    * self._lissa_batch_size
+                    * self.recur_depth
+                )
+            else:
+                assert isinstance(self.recur_depth, int)
+                recursion_depth = self.recur_depth
+
+            inv_hvp = self.get_inverse_hvp_lissa(
+                vs=test_grads,
+                model=self.model,
+                used_params=self._used_params,
+                lissa_dataloader=self._lissa_dataloader,
+                num_samples=self.num_samples,
+                recursion_depth=recursion_depth,
+                damping=self.damping,
+                scale=self.scale,
+            )
+
+            # Now, w.r.t. the current test examples, we iterate through the train set again to gain
+            # calculate the influence score for each training examples.
+            influences = torch.zeros(len(self._train_loader))
+            train_outputs = []
+            for train_idx, training_batch in enumerate(
+                tqdm(self._train_loader, desc="Train set index")
+            ):
+                self.model.train()
+                train_output_dict = self.model(**training_batch)
+                train_loss = train_output_dict["loss"]
+                train_outputs.append({"loss": train_loss.detach().item()})
+
+                self.model.zero_grad()
+                train_grads = autograd.grad(train_loss, self._used_params)
+                # dL_test * d theta as in 2.2 of [https://arxiv.org/pdf/2005.06676.pdf]
+                influences[train_idx] = torch.dot(inv_hvp, self.flatten_tensors(train_grads)).item()
+
+            # For each test example, we output top-k and training instances
+            # Then, we record the top-k training instances
+            _, indices = torch.topk(torch.tensor(influences), self._k)
+            top_k_train_instances: List[Instance] = []
+            for idx in indices:
+                train_instance = self._train_loader._instances[idx]
```

Comment:
Due to the declaration [here](https://github.com/allenai/allennlp/blob/59b92106443ff6ce26776beef5259dee3a422b40/allennlp/data/data_loaders/multiprocess_data_loader.py#L273), mypy will give 
```
allennlp/interpret/influence_interpreters/simple_influence.py:197: error: Value of type "Optional[List[Instance]]" is not indexable
```
Do you know how to fix it?

Reply:
I think it would be better to load the training instances into a list, and then you don't need to keep `_train_loader` around.

Reply:
Interesting point. Will do